src/tsolvers/egraph/EgraphSolver.cc

```diff
@@ -330,20 +332,24 @@ void Egraph::constructTerm(PTRef tr) {
 
         boolTermToERef.insert(tr_pure, er_pure);
         boolTermToERef.insert(tr_neg, er_neg);
-    }
-
-    else {
-        sym = enode_store.addSymb(tm.symb());
-        cdr = ERef_Nil;
-        for (int j = tm.size() - 1; j >= 0; j--) {
-            assert(enode_store.termToERef.has(tm[j])); // The child was not inserted
-            ERef car = enode_store.termToERef[tm[j]];
+    } else {
+
+        if (not logic.isNonBoolIte(tr)) {
+            sym = enode_store.addSymb(tm.symb());
+            cdr = ERef_Nil;
+            for (int j = tm.size() - 1; j >= 0; j--) {
+                assert(enode_store.termToERef.has(tm[j])); // The child was not inserted
+                ERef car = enode_store.termToERef[tm[j]];
 #ifdef VERBOSE_EUF
-            ERef prev_cdr = cdr;
-            assert (enode_store[car].getRoot() == car);
-            assert (enode_store[cdr].getRoot() == cdr);
+                ERef prev_cdr = cdr;
+                assert (enode_store[car].getRoot() == car);
+                assert (enode_store[cdr].getRoot() == cdr);
 #endif
-            cdr = enode_store.addList(car, cdr);
+                cdr = enode_store.addList(car, cdr);
+            }
+        } else {
+            sym = enode_store.addIteSymb(tr);
```

Comment:
I am just curious, why a separate (anonymous) symbol node is needed for each ITE term? Why a single symbol node cannot be re-used?

Reply:
The logic is that the ITEs are treated as variables, but an ITE with arguments $\bf{x}$ must be different from an ITE with arguments $\bf{y}$.  Therefore they must be different variables.   If a symbol node is reused, they become the same variable.

The symbol name, however, does not need to be different, since it's enough that the ERef of the symbol differs.

I'm not sure if this latter is confusing, and I could see it having some debugging benefits if each symbol would also be different, but since it's not necessary I didn't do it so far.

Reply:
I think I see it. I did not realize before that you are ignoring the arguments of the ITE. Now it makes sense, thanks!